utils/providers.py

```diff
@@ -147,6 +148,14 @@ def setup_a_provider(prov_class=None, prov_type=None, validate=True, check_exist
         providers = list_infra_providers()
 
     result = None
+
+    # If there is already a suitable provider, don't try to setup a new one.
+    already_existing = filter(is_provider_setup, providers)
+    if already_existing:
+        return setup_provider(  # This will run a refresh too
+            random.choice(already_existing), validate=validate, check_existing=check_existing)
+
+    # We need to setup a new one
     for provider in providers:
         try:
             result = setup_provider(provider, validate=validate, check_existing=check_existing)
```

Comment:
Oh but I see what you mean........If the _first_ one doesn't exist it will set it up.......